docs: update tutorial to use uv (#518)
UCL-ARC/python-tooling: tutorial.md

<details><summary>Click to expand... </summary> <!-- markdownlint-disable-line MD033 -->

> To start using the template you will need at least the following software tools installed:

Author: ```suggestion
To use the template you will need at least the following software tools installed
```

- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
- [uv](https://docs.astral.sh/uv/getting-started/installation/)

Additionally, [the GitHub command line interface (CLI)](https://cli.github.com/), [pre-commit](https://pre-commit.com/), and [tox](https://tox.wiki) are required to complete some of the follow on steps for using the package generated with the template, so while they are not strictly needed, you will get more out of the tutorial if you have them installed. For the follow on exercises you will also need an [account set up on GitHub](https://github.com/join) if you don't already have one.

> An easy way to install the software you need is by using `uv`, which can create a dedicated environment with the necessary packages installed using the step by step instructions below:

Author: ```suggestion
An easy way to install the software you need is by using `uv`, which can create a dedicated environment with the necessary packages installed using the step by step instructions below
```

> 1. Open a terminal window:

Author: ```suggestion
1. Open a terminal window
```

   - On Windows: open the Start menu from the taskbar, type `cmd` in the search field, then click [command prompt](https://en.wikipedia.org/wiki/Command-line_interface#Command_prompt) from the results.
   - On MacOS: click the Launchpad icon in the Dock, type `Terminal` in the search field, then click `Terminal` from the results.
   - On Linux: open the default terminal application installed in your distribution.

> 2. In the opened terminal window, use:

Author: ```suggestion
2. In the terminal window, use
```

   ```sh
   uv tool install cookiecutter gh pre-commit tox
   ```

   to install the "tools" necessary for creating and using a package using the `UCL-ARC/python-tooling` cookiecutter template.

> 3. To check that all the dependencies have installed correctly, run each of the following commands in turn, one at a time,

Author: ```suggestion
3. To check that all the dependencies have installed correctly, run each of the following commands in turn, one at a time
```

<details><summary>Click to expand... </summary> <!-- markdownlint-disable-line MD033 -->

To use the template you will need at least the following software tools installed

- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
- [uv](https://docs.astral.sh/uv/getting-started/installation/)
- [the GitHub command line interface (CLI)](https://github.com/cli/cli?tab=readme-ov-file#installation)

> Additionally, [pre-commit](https://pre-commit.com/) and [tox](https://tox.wiki) are required to complete some of the follow on steps for using the package generated with the template, so while they are not strictly needed, you will get more out of the tutorial if you have them installed. For the follow on exercises you will also need an [account set up on GitHub](https://github.com/join) if you don't already have one.

Reviewer: ```suggestion
Additionally, [pre-commit](https://pre-commit.com/) and [tox](https://tox.wiki) are required to complete some of the follow along steps for using the package generated with the template, so while they are not strictly needed, you will get more out of the tutorial if you have them installed. For the follow along exercises you will also need an [account set up on GitHub](https://github.com/join) if you don't already have one.
```

Reviewer 2: ["Follow on" is correct here for the sense of  "things following on from"](https://dictionary.cambridge.org/dictionary/english/follow-on) which is what I intended when I originally wrote this in the context of running this as a tutorial at the festival. "Follow along" would I think change the meaning

Reviewer: When reading it, I genuinely wasn't sure what was intended. If it's that use case then I think we should use following.

> An easy way to install the software you need is by using `uv`, which can create a dedicated environment with the necessary packages installed using the step by step instructions below

Reviewer 2: While I think `uv` can be used to handle the Python based tools needed (`cookiecutter`, `tox`, `pre-commit`) it doesn't provide a solution for the GitHub CLI or Git itself which would need to be separately installed (while `conda` does allow installing non Python dependencies). I don't think this is a blocker to recommending using `uv` but I think this section will need to be reworked slightly to indicate Git and GitHub CLI will need to be installed separately and with links to relevant installation instructions.

Author: Added a sentence ✅

1. Open a terminal window

   - On Windows: open the Start menu from the taskbar, type `cmd` in the search field, then click [command prompt](https://en.wikipedia.org/wiki/Command-line_interface#Command_prompt) from the results.
   - On MacOS: click the Launchpad icon in the Dock, type `Terminal` in the search field, then click `Terminal` from the results.
   - On Linux: open the default terminal application installed in your distribution.

> 2. In the terminal window, use

Reviewer: ```suggestion
2. In the terminal window, type
```

   ```sh
   uv tool install cookiecutter
   uv tool install pre-commit

>    uv tool install tox

Reviewer 3: Do they really need `tox` and `pre-commit`?

Author: We don't! Thanks!

>    uv tool install tox

Reviewer: Can these be one line?

Reviewer 2: Not with one `uv` command - `uv tool install --help` gives usage as

```
Usage: uv tool install [OPTIONS] <PACKAGE>
```

and trying with multiple package names gives an error.

It might be possible to stack multiple commands on a line using separators but I'm not sure if this would be easy to do in a way that will work across operating system / shells and also would probably be less readable.

Author: We don't need `pre-commit` and `tox` so that should solve it.

<details><summary>Click to expand... </summary> <!-- markdownlint-disable-line MD033 -->

To use the template you will need to install the following software tools

- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
- [uv](https://docs.astral.sh/uv/getting-started/installation/)
- [the GitHub command line interface (CLI)](https://github.com/cli/cli?tab=readme-ov-file#installation)

The instructions for installing these tools can be accessed by clicking the individual links above.

> Additionally, [cookiecutter](https://cookiecutter.readthedocs.io) will be required to generate your package from them template, which can be installed using the instructions below. For the follow on exercises you will also need an [account set up on GitHub](https://github.com/join) if you don't already have one.

Author: ```suggestion
Additionally, [cookiecutter](https://cookiecutter.readthedocs.io) will be required to generate your package from the template, which can be installed using the instructions below. For the follow on exercises you will also need an [account set up on GitHub](https://github.com/join) if you don't already have one.
```

We will first go through the steps for creating a new package using the `UCL-ARC/python-tooling` cookiecutter template.

> 1. Open a new terminal/command prompt window and change the working directory to the path you wish to create the package in using the `cd` (change directory) command.

Author: ```suggestion
1. Open a new terminal/command prompt window and change the working directory to the path you wish to create the package in using the `cd` (change directory) command
```

Once you are in your `{project_slug}` directory, a virtual environment can be created by running the following in a terminal/command prompt window

```sh
uv venv .venv
```

This will create a new environment with name `.venv`, which can be activated for installing packages using

```sh
source .venv/bin/activate
```

or if on Windows

```sh
.venv\Scripts\activate
```

You should see `(.venv)` on the left of your terminal window. Once you have activated the environment you can use the following command to install any packages within the `.venv` environment

```sh
uv pip install numpy
```

The environment can be deactivated using

```sh
deactivate
```

You can even use an arbitrary Python version to create `uv` environments, and `uv` will fetch that Python version if it is not available locally

```sh
uv venv --python 3.11.6
```

> Alternatively, you can use the [`venv` module](https://docs.python.org/3/library/venv.html), which is slower and has fewer features, when compared to `uv`, but is built-in to the Python standard library. `venv` has the advantage of being available in any Python (3.3+) environment, but unlike `uv` will not by itself allow you to use a different Python version from the system level install. A common pattern is to store the virtual environment files in a directory `.venv` within the root directory of the project repository. This can be achieved by running

Reviewer 3: I just ran through these for sanity as a fresh user in macOS.

The only stumbling point is from here on, it's not clear that we should preferentially be using UV.

That Is, if I skip or skim read the "Alternatively..." part and then try to `python -m pip install X` I will get:
```
(python-template) teaching@scafell python-template % python -m pip install numpy
/Users/teaching/python-template/.venv/bin/python: No module named pip
```
Obviously because I didn't install `pip` in my `venv`. But I _should_ be running `uv pip install` all the time.

Should we cull some of the information that follows? Or put it into a `<details></details>` block to hide it away?

I think we want to end the `</details>` block around
```
263 Once you have created and activated a virtual environment for the project, ...
```
Because the `uv` users (who will be in the majority?) will also need to install their own package into the `uv venv`.

Once you are in your `{project_slug}` directory, a virtual environment can be created by running the following in a terminal/command prompt window

```sh
uv venv .venv
```

This will create a new environment with name `.venv`, which can be activated for installing packages using

```sh
source .venv/bin/activate
```

or if on Windows

```sh
.venv\Scripts\activate
```

You should see `(.venv)` on the left of your terminal window. Once you have activated the environment you can use the following command to install any packages within the `.venv` environment

```sh
uv pip install numpy
```

The environment can be deactivated using

```sh
deactivate
```

You can even use an arbitrary Python version to create `uv` environments, and `uv` will fetch that Python version if it is not available locally

```sh
uv venv --python 3.11.6
```

> Alternatively, you can use the [`venv` module](https://docs.python.org/3/library/venv.html), which is slower, has less features, and is not developed rapidly when compared to `uv`, but is built-in to the Python standard library. Similar to `uv`, a virtual environment can be created using

Reviewer: ```suggestion
Alternatively, you can use the [`venv` module](https://docs.python.org/3/library/venv.html), which is slower, has fewer features, and is not developed rapidly when compared to `uv`, but is built-in to the Python standard library. Similar to `uv`, a virtual environment can be created using
```

> Alternatively, you can use the [`venv` module](https://docs.python.org/3/library/venv.html), which is slower, has less features, and is not developed rapidly when compared to `uv`, but is built-in to the Python standard library. Similar to `uv`, a virtual environment can be created using

Reviewer 2: I'd say that the specific point made previously that the built in `venv` module does not allow you to manage the base Python version used in the environment would be useful to keep here. I would also say that 'developed rapidly' is not necessarily always a benefit 🙃 particularly for infrequent / less experienced users for which a more stable tool might be better.

Author: Should be back now
